core/src/main/java/org/everit/json/schema/ObjectComparator.java

```diff
@@ -33,6 +45,15 @@ public static boolean deepEquals(Object obj1, Object obj2) {
                 return false;
             }
             return deepEqualObjects((JSONObject) obj1, (JSONObject) obj2);
+        } else if (obj1 instanceof Number) {
+            if (!(obj2 instanceof Number)) {
+                return false;
+            } else if (!obj1.getClass().equals(obj2.getClass())) {
+                String obj1AsString = obj1 instanceof Double && (Double) obj1 >= 10000000 ? numberFormat.format(obj1) : JSONObject.numberToString((Number) obj1);
+                String obj2AsString = obj2 instanceof Double && (Double) obj2 >= 10000000 ? numberFormat.format(obj2) : JSONObject.numberToString((Number) obj2);
+
```

Comment:
If the number is < 10000000 , then it gets cast twice (once to `Double`, then to `Number`). Please let's cast only once. Also please extract this string conversion to a separate method, because it is duplicated for `obj1` and `obj2`.

Why did you choose `10000000` as the treshold? Is there any specific reason?

Reply:
I chose 10000000 because the numbers larger than or equal to it are represented in scientific notation but numbers smaller than 0.001 are also represented in scientific notation those should be handled as well. Anyway I decided to implement a different approach because NumberFormat is determined by Locale. I could bind the format to a specific Locale but still there would be a small chance that the formatting behaves in an unexpected way in some special cases that I do not see now.